File: python/0143.py
from data_strcutres import ListNode
from collections import deque


class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        #  self.with_dqueue(head)
        self.reorder_after_middle(head)

    def reorder_after_middle(self, head: ListNode) -> None:
        if not head: return

        # find the middle (the slow cursor)
        fast = slow = head
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next

        # reorder the list after middle
        middle = slow
        if not middle.next: return
        cur = middle.next
        while cur.next:
            next = cur.next
            cur.next = next.next
            next.next = middle.next
            middle.next = next

        # merge forward and backward lists
        forward, backward = head, middle.next
        # cut off non-forward part in forward list
        middle.next = None
        # check backward only since it's shorter or equal to forward
        while backward:
            t1, t2 = forward.next, backward.next
            forward.next = backward
            backward.next = t1
            forward, backward = t1, t2


    def with_dqueue(self, head: ListNode) -> None:
        if not head: return

        vec = deque()
        cur = head
        while cur:
            vec.append(cur)
            cur = cur.next

        # link the first i-th element and last i-th element
        # in different direction interleavely.
        prev = vec.popleft()
        left = False
        while vec:
            if left:
                prev.next = vec.popleft()
            else:
                prev.next = vec.pop()
            prev = prev.next
            left = not left

        # trim the end to avoid cycle
        prev.next = None


a = ListNode(1)
#  a.next = ListNode(2)
#  a.next.next = ListNode(3)
#  a.next.next.next = ListNode(4)
#  a.next.next.next.next = ListNode(5)

Solution().reorderList(a)

File: cpp/0143.cpp
#include <vector>
#include <string>
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v) {}
    string into_string() {
        string ctx = to_string(this->val);
        ListNode* curr = this->next;
        while (curr) {
            ctx += "->" + to_string(curr->val);
            curr = curr->next;
        }
        return ctx;
    }
};

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* preMid = findPreMidium(head);
        ListNode* half = preMid->next;
        preMid->next = nullptr;
        head = mergeTwoLists(head, reverseList(half));
    }

private:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) return head;
        ListNode* root = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return root;
    }

    ListNode* findPreMidium(ListNode* head) {
        if (!head) return nullptr;
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }

    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (!l1) return l2;
        if (!l2) return l1;
        l1->next = mergeTwoLists(l2, l1->next);
        return l1;
    }

};

ListNode* vecToList(vector<int> vec) {
    ListNode* dummy = new ListNode(0);
    ListNode* curr = dummy;
    for (auto v: vec) {
        curr->next = new ListNode(v);
        curr = curr->next;
    }
    return dummy->next;
}

int main() {
    Solution sol;
    vector<int> e = {0, 2, 4, 6, 8};
    vector<int> o = {1, 3, 5, 7, 9, 11};
    ListNode* even = vecToList(e);
    ListNode* odd = vecToList(o);
    sol.reorderList(odd);
    cout << "Even: " << even->into_string() << endl;
    cout << "Odd: " << odd->into_string() << endl;
    // cout << "Merge: " << mergeTwoLists(even, odd)->into_string() << endl;
    // cout << "Mid of even: " << findMidium(even)->into_string() << endl;
    // cout << "Mid of odd: " << findMidium(odd)->into_string() << endl;
    // cout << "Reverse of even: " << reverseList(even)->into_string() << endl;
    // cout << "Reverse of odd: " << reverseList(odd)->into_string() << endl;
}
